lib/portlayer/exec/base.go

```diff
@@ -461,14 +454,77 @@ func (c *containerBase) waitForSession(ctx context.Context, id string) error {
 	return c.waitFor(ctx, key)
 }
 
-func (c *containerBase) waitForExec(op trace.Operation, id string) error {
+func (c *containerBase) WaitForExecStart(op trace.Operation, id string) error {
 	defer trace.End(trace.Begin(id, op))
 
 	// guestinfo key that we want to wait for
 	key := extraconfig.CalculateKeys(c.ExecConfig, fmt.Sprintf("Execs.%s.Started", id), "")[0]
 	return c.waitFor(op, key)
 }
 
+// Waits for the exec'd task to complete.
+//TODO: we really need structured Task style errors so that we can be more explicit in our handling upstream.
+func (c *containerBase) WaitForExec(op trace.Operation, id string) error {
```

Comment:
We spoke directly about this, but here for posterity...

I think altering this so that we're just waiting for a change in a specified set of keys will be the most effective way to implement what we're after (`WaitForKeyChange(op trace.Operation, currentKeyValues map[string]string)`  maybe).

I'd also request a `containerBase.TaskWait()` or similar function that does the calls to extraconfig.CalculateLeys before calling the general function to wait for the change. This is so that the `task` package doesn't start depending on the implementation of the config store.